File: ruby/gameboard.rb
require_relative "hangman"

class Gameboard
    attr_accessor :random_word, :guessed_chars, :num_wrong, :placeholder, :game_over
    
    def initialize() 
        @random_word = choose_random_word()
        @guessed_chars = []
        @num_wrong = 0
        @placeholder = ""
        @game_over = false
    end

    def choose_random_word() 
        lines = File.read("../words.txt").split("\n")
        @random_word = lines.sample()
    end

    def choose_game_image()
        case @num_wrong
        when 0
            Hangman::EMPTY
        when 1
            Hangman::HEAD
        when 2
            Hangman::NECK
        when 3
            Hangman::ONE_ARM
        when 4
            Hangman::TWO_ARMS
        when 5 
            Hangman::ONE_LEG
        when 6 
            Hangman::FINISHED_MAN
        else 
            Hangman::DEAD_MAN
        end
    end 

    def choose_letter() 
        input = gets.chomp
        begin
            input = input[0]
            input = input.downcase
            unless @guessed_chars.include?(input)
                @guessed_chars.push(input) 
            else
                puts "Already guessed that letter!"    
            end
        rescue
            puts "Invalid input! Defaulting to 'a'"
            unless @guessed_chars.include?('a')
                @guessed_chars.push('a')
            end
        end 
    end

    def update_placeholder()
        @placeholder = ""
        @random_word.length.times do |i|
            @placeholder += @guessed_chars.include?(@random_word[i]) ? 
                @random_word[i] + " " :  "_ " 
        end
        @placeholder = @placeholder.chomp
    end

    def calculate_num_wrong()
        correct_guesses = @guessed_chars.select do |c| 
            random_word.include?(c) 
        end 
        @num_wrong = @guessed_chars.length - correct_guesses.length
    end

    def display()
        image = choose_game_image()
        puts image + "\n" + @placeholder + "\n"
    end

    def check_game_over()
        correct_guesses = @guessed_chars.select do |c| 
            random_word.include?(c) 
        end 
        random_word_length = @random_word.split("").uniq.join("").length
        if @num_wrong >= 7 
            puts "Game over! The word was #{@random_word}"
            @game_over = true
        elsif correct_guesses.length >= random_word_length
            puts "You win! The word was #{@random_word}"
            @game_over = true
        else 
            @game_over = false
        end
    end
end

File: elixir_app/lib/gameboard.ex
defmodule GameBoard do
    def run(pid) do
        num_wrong = GameState.check_wrong_guesses(pid)
        IO.puts GameBoard.select_game_image(num_wrong)

        guess = IO.gets(:stdio, "Please enter a letter: ") |> String.trim()
        guess =
            case guess do
                "" ->
                    IO.puts "Invalid input. Defaulting to 'a'"
                    "a"
                c when c != "" -> 
                    String.at(c, 0) |> String.downcase
            end

        temp_len = Enum.count(GameState.get_guesses(pid))
        GameState.guess_letter(pid, guess)
        new_len = Enum.count(GameState.get_guesses(pid))

        if temp_len != new_len do
            GameState.check_last_guess(pid)
        end

        game_over = GameState.check_game_over(pid)
        if game_over == false do
            run(pid)
        end
    end

    def choose_random_word() do
        {err, file} = File.read("../words.txt")
        case err do
            :ok ->
                lst = String.split(file)
                len = Enum.count(lst)
                num = Enum.random(0..len)
                Enum.at(lst, num)
            _ ->
                "Error reading file"
        end
    end

    def select_game_image(num_wrong) do
        case num_wrong do
            0 -> Hangman.empty()
            1 -> Hangman.head()
            2 -> Hangman.neck()
            3 -> Hangman.one_arm()
            4 -> Hangman.two_arms()
            5 -> Hangman.one_leg()
            6 -> Hangman.finished_man()
            _ -> Hangman.dead_man()
        end
    end
end
